src/scp/QuorumIntersectionChecker.cpp

```diff
@@ -0,0 +1,1184 @@
+//         intersect by assumption.
+//
+// So what is this good for? If we treat our problem as searching for minqs
+// rather than more general quorums, we can do an early return in the
+// enumeration any time we encounter any quorum at all: we're growing bottom-up
+// from sets to supersets, but no superset of a quorum is a minq, so once we're
+// at a committed set C that's a quorum there's no need to look in any further
+// sub-branches of the search enlarging C: there are no more minqs down those
+// branches.
+//
+// So we test for quorum-ness, and then for minq-ness (checking the
+// complement for a non-intersecting quorum if so), and then return early in
+// either case:
+//
+//     def enumerate(C, R):
+//         return if #C > MAXSZ
+//         if C is a quorum:
+//             fail if C is a minq and (N \ C) has a quorum
+//             return                                         ← new early exit
+//         return if R = ∅
+//         return if (C ∪ R) has no quorum or
+//             its maximal quorum isn't a superset of C
+//         pick some node Nᵢ ∈ R
+//         call enumerate(C ∪ Nᵢ, R-Nᵢ)
+//         call enumerate(C, R-Nᵢ)
+//
+//
+// Refinement 7: heuristic node selection
+// ======================================
+//
+// Observe above that we pick "some node Nᵢ ∈ R" but don't say how. It turns out
+// that the node we choose has a dramatic effect on the amount of search space
+// explored. Picking right means we get to an early exit (or failure) ASAP;
+// picking wrong means we double the search without hitting any early exits in
+// the next layer of the recursion.
+//
+// There's no good single answer for the best next-node to divide on; probably
+// knowing this would be equivalent to solving the problem anyways. But we can
+// use some heuristics. I've explored several and not come up with any better
+// (empirically) than in Lachowski's initial code, so we stick with that
+// heuristic here: we pick the node with the highest indegree, when looking at
+// the subgraph of the remaining nodes R and edges between them. If there are
+// multiple such nodes we pick among them randomly. This seems to favor
+// discovering participants in quorums early, taking the quorum-related early
+// exits ASAP.
+//
+//
+// Refinement 8: Time for some graph theory
+// ========================================
+//
+// There is one last step in the development here and it also involves minqs but
+// not early exits in the enumeration as such. The enumerate function above is
+// left alone, but we reduce the inputs to it from minqs in P(N) to minqs in
+// P(M) for some M ⊆ N, reducing our search a lot. For this we need graph
+// theory!
+//
+// The trick here is to recognize that there's a relationship between minqs and
+// strongly connected components (SCCs) in the induced directed graph
+// Gₙ. Specifically that the following theorem holds:
+//
+//     Any minq of N is entirely contained in an SCC of the induced graph Gₙ.
```

Comment:
Oh, well the simple answer isn't me being clever or anything, I just that I did the implementation & explanation from @fixxxedpoint's original post in galactictalk where he mentioned the weaker property of being-contained-in an SCC: https://galactictalk.org/d/1613-quorum-intersection-a-tool-for-verifying-quorum-intersection-property

I mean, every subset of an SCC is strongly connected. So in some sense it's not saying anything different.

But reading this and thinking about it more: I think it might be a bit _misleading_ to the reader to identify minqs and SCCs, because typically (though perhaps not exclusively?) "SCC" -- not just strongly connected sets, but strongly connected _components_ -- refers to the partition of a graph in which the set is both strongly connected and _maximal_, whereas not every minq is such a _maximal_ strongly connected set: they are usually contained in a larger strongly-connected set (identified in corollary 4). Indeed, we only enumerate the powerset of a single maximal SCC looking for its minqs, and we tend to find lots of minqs inside that one SCC. So I think the reader would be a bit confused if we said "an SCC == a minq"; this code uses the term "SCC" to refer to those maximal SCCs found by Tarjan's algorithm that partitions the graph, rather than the (strongly connected) minqs inside of that SCC.

Clear enough? I hope I am making sense, this is at least how I read the divergence in terminology. Might be worth asking @fixxxedpoint specifically!

Reply:
(resolved in discussion)